_source/_posts/2024-04-30-express-universal-logout.md

```diff
@@ -0,0 +1,674 @@
+
+### Sign a user out of the Todo app
+
+Under the `apps/todo-app/src/app/components/todolist.tsx` file. Find the function for creating a new task onNewTask function and add the following code to catch a 401 error and redirect them back to the login page.
+
+```ts
+ if (!res.ok)
+{if (res.status === 401) {
+// Redirect user back to login
+window.location.href = '/';
+        } else {
+          // Handle other errors
+          throw new Error('Error occurred while fetching data');
+        }}     
+
+The onNewTask function will now look like this:
+
+import { useEffect, useState } from 'react';
+import { useAuthState } from './authState';
+
+interface ITodo {
+  id: number;
+  task: string;
+  completed: boolean;
+}
+
+export const Todos = () => {
+  const [todoList, setTodoList] = useState<ITodo[]>([]);
+  const [newTask, setNewTask] = useState<string>('');
+  const { authState } = useAuthState();
+
+  const API_BASE_URL = '/api/todos';
+
+  const onNewTask = () => { 
+    const apiCall = async () => {
+      try {
+        const res = await fetch(API_BASE_URL, {
+          method: 'POST',
+          credentials: 'same-origin',
+          mode: 'same-origin',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify({ task: newTask })
+        });
+
+        if (!res.ok){if (res.status === 401) {
+// Redirect user back to login
+window.location.href = '/';
+        } else {
+          // Handle other errors
+          throw new Error('Error occurred while fetching data');
+        }}
+
+        const todo = await res.json();
+        setTodoList([...todoList, todo]);
+        setNewTask('');
+      } catch (error: unknown) {
+        console.error(error);
+      }
+    };
+    apiCall();
+  };
+
+```
+
+>**Checkpoint**: Let's retest, ending Trinity's session, and watch what happens when you attempt to add a new task to the Todo app.
+
+>**Improve your code**: Notice the code above only handles a 401 response from the server when adding a new task. How might you handle 401 errors globally? You can use fetch or [Axios Interceptor](https://axios-http.com/docs/interceptors). The completed workshop code handles this using fetch; check it out here [Universal Logout Workshop Complete](https://github.com/oktadev/okta-enterprise-ready-workshops/blob/ul-workshop-complete/apps/todo-app/src/app/components/useTodoApi.tsx).
+
+### Revoke a user's tokens
+This web application architecture used cookie-based sessions instead of session tokens to authenticate to the backend resources. However, in the case of mobile apps and single-page applications, you will need to revoke refresh tokens on the frontend. As per the [spec](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-global-token-revocation#name-revocation-expectations), written by [Aaron Perecki](https://aaronparecki.com/) a successful logout will require revoking a user's refresh token. 
```

Comment:
```suggestion
This web application architecture uses cookie-based sessions instead of session tokens to authenticate to the backend resources. However, in the case of mobile apps and single-page applications, you'll need to revoke refresh tokens on the front end. As per the [spec](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-global-token-revocation#name-revocation-expectations), written by [Aaron Perecki](https://aaronparecki.com/) a successful sign-out will require revoking a user's refresh token. 
```